### DISCUSSION TITLE
How to navigate to other page outside of Component?

### DISCUSSION THREAD

**Discussion by JesuisTong** [vercel/next.js#60457] [App Router] [answered]
Imagine the following scenario:
Client establish a websocket, client should navigate to page received from websocket message.
Code below:
```js
// ws.js
"use client";
const ws = new Websocket('ws://xxx');
ws.onmessage = e => {
  // navigate logic, sth like router.push(e.message);
}
```
But [the official document](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating) told that there are only two way to navigate: `<Link>`,`useRouter`. It's so hard to navigate when some logic must do like this.
Is there other way to navigate?

**Comment by icyJoseph** [comment]
Since there's no Router singleton to invoke the push/replace, methods on, you could either:

- use window.location, https://developer.mozilla.org/en-US/docs/Web/API/Window/location#example_1_navigate_to_a_new_page
- if you still want to be within React's scope, put the logic inside a `useEffect`, and there you'll have access to the Router push method.

**Reply by JesuisTong** [reply]
Thanks, your ideas are worked,
but the first one will break UX, the second one will merge many logic into UI component or hooks, that may not be elegant.
Could i store `useRouter()` as a global variable?

**Reply by icyJoseph** [reply, marked as answer]
Well, I'd say that as long as the framework doesn't expose a hook to globally do this, (which I've seen being request before), you are a bit out of luck as far as having a clean solution, and variations on the second approach would be the most elegant.

Nothing prevents you from, doing something like, storing a reference to `router.push` at the top of a component file, and such, but that'll be prone to bugs, and hard to maintain in my opinion.

Alternatively, you could, create a function that wraps your websocket logic, so that it takes a dispatcher/callback (router.push/replace) as input, and then you'd call that function within a use-effect, which would still be correct, since use-effect exists to synchronize with external stuff.

So to avoid having to:

> merge many logic into UI component or hooks

You should try to use use-effect, as a way to synchronize with the websockets.

**Reply by icyJoseph** [reply]
The problem here is something you might have seen in Redux before, if you had used that. How do we make dispatch available outside of React tree? In this case it is even more tricky, because the new Router uses `useReducer`, and that binds it a bit too tight to the React tree. You can even see the App Router events in the Redux Dev Tools Browser Extension.

**Reply by JesuisTong** [reply]
Thanks for your meticulous explains!
Agree with you that i should use `useRouter` to do navigate, but also i want to navigate in pure js function.
So i have an idea below:
```js
// the common app layout
export default () => {
  const router = useRouter();
  useEffect(() => {
    routerSubscribe.on(pathname => router.push(pathname));
    return () => routerSubscribe.off();
  }, [router]);
};

// non react js function scope
const callback = () => {
  // logic
  // navigate
  routerSubscribe.emit('/a/b/c');
};
```
Is it a good way to solve this?

**Reply by icyJoseph** [reply]
Yeah, something like that should do the trick.
